File: test/language/array_types/cpp/ImplicitArrayInt24Test.cpp
#include "gtest/gtest.h"

#include "array_types/implicit_array_int24/ImplicitArray.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

namespace array_types
{
namespace implicit_array_int24
{

class ImplicitArrayInt24Test : public ::testing::Test
{
protected:
    void writeImplicitArrayToByteArray(zserio::BitStreamWriter& writer, size_t numElements)
    {
        for (size_t i = 0; i < numElements; ++i)
            writer.writeSignedBits(static_cast<int32_t>(i), 24);
    }
};

TEST_F(ImplicitArrayInt24Test, bitSizeOf)
{
    const size_t numElements = 55;
    std::vector<int32_t> array;
    array.reserve(numElements);
    for (size_t i = 0; i < numElements; ++i)
        array.push_back(static_cast<uint8_t>(i));
    ImplicitArray implicitArray;
    implicitArray.setArray(array);

    const size_t bitPosition = 2;
    ASSERT_EQ(numElements * 24, implicitArray.bitSizeOf(bitPosition));
}

TEST_F(ImplicitArrayInt24Test, initializeOffsets)
{
    const size_t numElements = 55;
    std::vector<int32_t> array;
    array.reserve(numElements);
    for (size_t i = 0; i < numElements; ++i)
        array.push_back(static_cast<uint8_t>(i));
    ImplicitArray implicitArray;
    implicitArray.setArray(array);

    const size_t bitPosition = 2;
    ASSERT_EQ(bitPosition + numElements * 24, implicitArray.initializeOffsets(bitPosition));
}

TEST_F(ImplicitArrayInt24Test, read)
{
    const size_t numElements = 99;
    zserio::BitStreamWriter writer;
    writeImplicitArrayToByteArray(writer, numElements);
    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
    ImplicitArray implicitArray(reader);

    const std::vector<int32_t>& array = implicitArray.getArray();
    ASSERT_EQ(numElements, array.size());
    for (size_t i = 0; i < numElements; ++i)
        ASSERT_EQ(static_cast<int32_t>(i), array[i]);
}

TEST_F(ImplicitArrayInt24Test, write)
{
    const size_t numElements = 55;
    std::vector<int32_t> array;
    array.reserve(numElements);
    for (size_t i = 0; i < numElements; ++i)
        array.push_back(static_cast<uint8_t>(i));
    ImplicitArray implicitArray;
    implicitArray.setArray(array);

    zserio::BitStreamWriter writer;
    implicitArray.write(writer);

    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
    ImplicitArray readImplicitArray(reader);
    const std::vector<int32_t>& readArray = readImplicitArray.getArray();
    ASSERT_EQ(numElements, readArray.size());
    for (size_t i = 0; i < numElements; ++i)
        ASSERT_EQ(static_cast<int32_t>(i), readArray[i]);
}

} // namespace implicit_array_bit8
} // namespace array_types

File: test/language/array_types/java/array_types/ImplicitArrayInt24Test.java
package array_types;

import static org.junit.Assert.*;

import java.io.IOException;
import java.io.File;

import org.junit.Test;

import array_types.implicit_array_int24.ImplicitArray;

import zserio.runtime.ZserioError;
import zserio.runtime.array.IntArray;
import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.BitStreamWriter;
import zserio.runtime.io.FileBitStreamReader;
import zserio.runtime.io.FileBitStreamWriter;

public class ImplicitArrayInt24Test
{
    @Test
    public void bitSizeOf() throws IOException, ZserioError
    {
        final int numElements = 44;
        IntArray array = new IntArray(numElements);
        for (int i = 0; i < numElements; ++i)
            array.setElementAt(i, i);

        final ImplicitArray implicitArray = new ImplicitArray(array);
        final int bitPosition = 2;
        final int implicitArrayBitSize = numElements * 24;
        assertEquals(implicitArrayBitSize, implicitArray.bitSizeOf(bitPosition));
    }

    @Test
    public void initializeOffsets() throws IOException, ZserioError
    {
        final int numElements = 66;
        IntArray array = new IntArray(numElements);
        for (int i = 0; i < numElements; ++i)
            array.setElementAt(i, i);

        final ImplicitArray implicitArray = new ImplicitArray(array);
        final int bitPosition = 2;
        final int expectedEndBitPosition = bitPosition + numElements * 24;
        assertEquals(expectedEndBitPosition, implicitArray.initializeOffsets(bitPosition));
    }

    @Test
    public void read() throws IOException, ZserioError
    {
        final File file = new File("test.bin");
        final int numElements = 99;
        writeImplicitArrayToFile(file, numElements);
        final BitStreamReader stream = new FileBitStreamReader(file);
        final ImplicitArray implicitArray = new ImplicitArray(stream);
        stream.close();

        final IntArray array = implicitArray.getArray();
        assertEquals(numElements, array.length());
        for (int i = 0; i < numElements; ++i)
            assertEquals(i, array.elementAt(i));
    }

    @Test
    public void write() throws IOException, ZserioError
    {
        final int numElements = 55;
        IntArray array = new IntArray(numElements);
        for (int i = 0; i < numElements; ++i)
            array.setElementAt(i, i);

        ImplicitArray implicitArray = new ImplicitArray(array);
        final File file = new File("test.bin");
        final BitStreamWriter writer = new FileBitStreamWriter(file);
        implicitArray.write(writer);
        writer.close();

        final ImplicitArray readImplicitArray = new ImplicitArray(file);
        final IntArray readArray = readImplicitArray.getArray();
        assertEquals(numElements, readArray.length());
        for (int i = 0; i < numElements; ++i)
            assertEquals(i, readArray.elementAt(i));
    }

    private void writeImplicitArrayToFile(File file, int numElements) throws IOException
    {
        final FileBitStreamWriter writer = new FileBitStreamWriter(file);

        for (int i = 0; i < numElements; ++i)
            writer.writeSignedBits(i, 24);

        writer.close();
    }
}
